iceoryx_dds/include/iceoryx_dds/internal/gateway/dds_to_iox.inl

```diff
@@ -23,26 +23,99 @@ namespace iox
 {
 namespace dds
 {
-template <typename channel_t>
-inline DDS2IceoryxGateway<channel_t>::DDS2IceoryxGateway() noexcept
-    : iox::dds::DDSGatewayGeneric<channel_t>()
+template <typename channel_t, typename gateway_t>
+inline DDS2IceoryxGateway<channel_t, gateway_t>::DDS2IceoryxGateway() noexcept
+    : gateway_t()
 {
 }
 
-template <typename channel_t>
-inline void DDS2IceoryxGateway<channel_t>::loadConfiguration(const GatewayConfig& config) noexcept
+template <typename channel_t, typename gateway_t>
+inline void DDS2IceoryxGateway<channel_t, gateway_t>::loadConfiguration(const GatewayConfig& config) noexcept
 {
+    iox::LogDebug() << "[DDS2IceoryxGateway] Configuring gateway.";
+    for (const auto& service : config.m_configuredServices)
+    {
+        if (!this->findChannel(service.m_serviceDescription).has_value())
+        {
+            setupChannel(service.m_serviceDescription, service.m_dataSize);
+        }
+    }
 }
 
+template <typename channel_t, typename gateway_t>
+inline void DDS2IceoryxGateway<channel_t, gateway_t>::discover(const iox::capro::CaproMessage& msg) noexcept
+{
+    /// @note not implemented - requires dds discovery which is currently not implemented in the used dds stack.
+}
 
-template <typename channel_t>
-inline void DDS2IceoryxGateway<channel_t>::discover(const iox::capro::CaproMessage& msg) noexcept
+template <typename channel_t, typename gateway_t>
+inline void DDS2IceoryxGateway<channel_t, gateway_t>::forward(const channel_t& channel) noexcept
 {
+    // Don't forward across channels that don't provide their size.
+    // Data size is required to determine the required memchunk size.
+    auto dataSize = channel.getDataSize();
```

Comment:
Having to know the data size in advance is somehow a limitation or? Why we nee this. Can you save a copy with this? Why not receive a sample from dds, check its size and then allocate in iceroyx and copy to the iceoryx chunk

Reply:
This design came from trying to keep iceoryx and dds loosely coupled.
The DDS abstraction API (CycloneDataReader) currently doesn't use any iceoryx-specific types. Instead it takes an arbitrary byte buffer as input. 
This buffer needs to be big enough for the sample, therefore the gateway, which is creating the buffer before the read, needs to know how big the buffer needs to be before doing the read.
I'm not sure why I prioritised this point so high, my thought process naturally went down that path for some reason.

The API could be updated to directly take an iceoryx publisher as input, then the CycloneDataReader could allocate the chunks itself.
Alternatively, I guess the read could be done in two calls: one to read a sample and check how large it is, and a second to copy the sample into the memory chunk.

Both options are pretty trivial to implement, I am leaning towards the first one, what do you think ?

Is data published for a particular iceoryx service description expected to vary in size ?

Reply:
FYI the implementation has been updated to not require the size ahead of time. Changes should be visible here soon.